{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "#input local file path to CRyPTIC tables\n",
    "file_path='../../cryptic_data_analysis/data_tables/cryptic-analysis-group'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "#read in files\n",
    "genomes = pd.read_pickle(file_path+\"/GENOMES.pkl.gz\")\n",
    "genomes.reset_index(level=0, inplace=True)\n",
    "genomes=genomes.loc[genomes['BELONGS_GPI']==True]\n",
    "\n",
    "phenotypes = pd.read_pickle(file_path+\"/UKMYC_PHENOTYPES.pkl.gz\")\n",
    "phenotypes.reset_index(level=1, inplace=True)\n",
    "phenotypes.reset_index(level=0, inplace=True)\n",
    "phenotypes=phenotypes.loc[phenotypes['BELONGS_GPI']==True]\n",
    "phenotypes.loc[((phenotypes['DRUG'] =='INH')&(phenotypes['BINARY_PHENOTYPE']=='I')), 'BINARY_PHENOTYPE'] = 'R'\n",
    "\n",
    "\n",
    "fq_phenotypes=phenotypes.loc[phenotypes.DRUG.isin(['LEV','MXF'])]\n",
    "\n",
    "samples = pd.read_pickle(file_path+\"/SAMPLES.pkl.gz\")\n",
    "samples.reset_index(level=2, inplace=True)\n",
    "samples.reset_index(level=1, inplace=True)\n",
    "samples.reset_index(level=0, inplace=True)\n",
    "\n",
    "mutations=pd.read_pickle(file_path+\"/MUTATIONS_GPI.pkl.gz\")\n",
    "mutations.reset_index(inplace=True)\n",
    "\n",
    "hets=pd.read_csv('het_mutations_dp2.csv')\n",
    "\n",
    "features = pd.read_csv(\"STRUCTURE_BASED_FEATURES_ALL.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#define list of lineage mutations not to include\n",
    "#ref for lineage mutations: Miotto ERJ 2017\n",
    "lin_muts=['gyrAS95T','gyrAE21Q','gyrAT80A','gyrAA90G','gyrAG247S','gyrAA384V',\n",
    "             'gyrAG668D','gyrAL712V', 'gyrAS250A','gyrAR252L','gyrAL398F','gyrAA463S','gyrAD639A','gyrAV742L',\n",
    "              'gyrBV301L','gyrBM291I','gyrBA403S']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "INH_AND_RIF_S    5788\n",
       "MDR              4353\n",
       "INH_MONOR        1470\n",
       "UNKNOWN           445\n",
       "RIF_MONOR         302\n",
       "Name: BACKGROUND, dtype: int64"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#create dataframe of RIF and INH resistant background for each isolate\n",
    "uid=[]\n",
    "background=[]\n",
    "for i in phenotypes.UNIQUEID.unique():\n",
    "    uid.append(i)\n",
    "    df=phenotypes.loc[phenotypes.UNIQUEID==i]\n",
    "    if len(df.loc[((df.DRUG=='RIF')&(df.BINARY_PHENOTYPE=='R'))|((df.DRUG=='INH')&(df.BINARY_PHENOTYPE=='R'))])==2:\n",
    "        background.append('MDR')\n",
    "    elif len(df.loc[((df.DRUG=='RIF')&(df.BINARY_PHENOTYPE=='R'))])==1 and len(df.loc[((df.DRUG=='INH')&(df.BINARY_PHENOTYPE=='S'))])==1:\n",
    "        background.append('RIF_MONOR')\n",
    "    elif len(df.loc[((df.DRUG=='RIF')&(df.BINARY_PHENOTYPE=='S'))])==1 and len(df.loc[((df.DRUG=='INH')&(df.BINARY_PHENOTYPE=='R'))])==1:\n",
    "        background.append('INH_MONOR')\n",
    "    elif len(df.loc[((df.DRUG=='RIF')&(df.BINARY_PHENOTYPE=='S'))])==1 and len(df.loc[((df.DRUG=='INH')&(df.BINARY_PHENOTYPE=='S'))])==1:\n",
    "        background.append('INH_AND_RIF_S')\n",
    "    else:\n",
    "        background.append('UNKNOWN')\n",
    "background_df=pd.DataFrame(list(zip(uid,background)), columns=['UNIQUEID', 'BACKGROUND'])\n",
    "\n",
    "background_df.BACKGROUND.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-6-9550fe76ecfb>:4: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  gyrase_mutations[\"GENEMUT\"]=gyrase_mutations['GENE'].astype('str')+gyrase_mutations['MUTATION']\n"
     ]
    }
   ],
   "source": [
    "#filter mutations so only have mutations in dna gyrase, label as homogeneous (HET=0)\n",
    "gyrase_mutations=mutations.loc[mutations.GENE.isin(['gyrA','gyrB'])&(mutations.IS_SYNONYMOUS==False)&(mutations.IN_PROMOTER==False)&(mutations.IS_SNP==True)&(mutations.IS_NULL==False)&(mutations.IS_FILTER_PASS==True)]\n",
    "\n",
    "\n",
    "gyrase_mutations[\"GENEMUT\"]=gyrase_mutations['GENE'].astype('str')+gyrase_mutations['MUTATION']\n",
    "gyrase_mutations = gyrase_mutations.loc[(-gyrase_mutations.GENEMUT.isin(lin_muts))]\n",
    "gyrase_mutations['IS_HET']=0\n",
    "\n",
    "gyrase_mutations = gyrase_mutations[['UNIQUEID','GENE', 'MUTATION', 'POSITION', \"GENEMUT\", \"IS_HET\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "#filter gyrase mutations with FRS <0.9 (remove nsynonymous muts and stop codons) and create complimentary dataframe to homogeneous mutations\n",
    "uid=[]\n",
    "gene=[]\n",
    "mutation=[]\n",
    "position=[]\n",
    "genemut=[]\n",
    "is_het=[]\n",
    "\n",
    "for index, row in hets.iterrows():\n",
    "\n",
    "    if str(row['HET_MUTATION'])!= 'nan' and str(row['HET_MUTATION'])[4]!= str(row['HET_MUTATION'])[-1] and str(row['HET_MUTATION'])[-1]!='O' and str(row['HET_MUTATION'])[5]!='-' and str(row['HET_MUTATION'])[-1]!='X' and str(row['HET_MUTATION'])[-4:] != 'STOP' and str(row['HET_MUTATION'])[4]!='!' and str(row['HET_MUTATION'])[4:8] != 'STOP':\n",
    "            uid.append(row['UNIQUEID'])\n",
    "            gene.append(str(row['HET_MUTATION'])[0:4])\n",
    "            mutation.append(str(row['HET_MUTATION'])[4:])\n",
    "            position.append(int(str(row['HET_MUTATION'])[5:-1]))\n",
    "            genemut.append(str(row['HET_MUTATION']))\n",
    "            is_het.append(1)\n",
    "            \n",
    "            \n",
    "\n",
    "new_hets=pd.DataFrame(list(zip(uid, gene, mutation, position, genemut, is_het)), \n",
    "                      columns=['UNIQUEID','GENE','MUTATION','POSITION', 'GENEMUT','IS_HET'])\n",
    "\n",
    "#remove lineage specific mutations\n",
    "new_hets=new_hets.loc[(-new_hets.GENEMUT.isin(lin_muts))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "#combine homogenous mutations and mutations with FRS <0.9\n",
    "gyrase_mutations=gyrase_mutations.append(new_hets)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "#filter to remove low quality phenotpyes and remove unecessary columns\n",
    "fq_phenotypes = fq_phenotypes.loc[fq_phenotypes.PHENOTYPE_QUALITY!='LOW']\n",
    "fq_phenotypes = fq_phenotypes[['UNIQUEID', 'DRUG','MIC', 'LOG2MIC','BINARY_PHENOTYPE', 'PHENOTYPE_QUALITY']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UNIQUEID</th>\n",
       "      <th>DRUG</th>\n",
       "      <th>MIC</th>\n",
       "      <th>LOG2MIC</th>\n",
       "      <th>BINARY_PHENOTYPE</th>\n",
       "      <th>PHENOTYPE_QUALITY</th>\n",
       "      <th>LINEAGE_NAME</th>\n",
       "      <th>COUNTRY_WHERE_SAMPLE_TAKEN</th>\n",
       "      <th>BACKGROUND</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>site.06.subj.06TB_1032.lab.06MIL2037.iso.1</td>\n",
       "      <td>MXF</td>\n",
       "      <td>0.25</td>\n",
       "      <td>-2.00</td>\n",
       "      <td>S</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>Lineage 4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>INH_MONOR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>site.06.subj.06TB_0404.lab.06MIL1373.iso.1</td>\n",
       "      <td>MXF</td>\n",
       "      <td>0.25</td>\n",
       "      <td>-2.00</td>\n",
       "      <td>S</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>Lineage 4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>INH_AND_RIF_S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>site.06.subj.SSM_0197-R.lab.06MIL0117.iso.1</td>\n",
       "      <td>MXF</td>\n",
       "      <td>&lt;=0.06</td>\n",
       "      <td>-4.06</td>\n",
       "      <td>S</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>Lineage 3</td>\n",
       "      <td>PAK</td>\n",
       "      <td>INH_AND_RIF_S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>site.04.subj.01328.lab.722926.iso.1</td>\n",
       "      <td>LEV</td>\n",
       "      <td>0.5</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>S</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>Lineage 2</td>\n",
       "      <td>IND</td>\n",
       "      <td>INH_AND_RIF_S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>site.03.subj.T1069.lab.T1069.iso.1</td>\n",
       "      <td>LEV</td>\n",
       "      <td>0.5</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>S</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>Lineage 4</td>\n",
       "      <td>TJK</td>\n",
       "      <td>RIF_MONOR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17598</th>\n",
       "      <td>site.05.subj.CA-1473.lab.CO-03323-19.iso.1</td>\n",
       "      <td>MXF</td>\n",
       "      <td>0.5</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>S</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>Lineage 4</td>\n",
       "      <td>PER</td>\n",
       "      <td>INH_AND_RIF_S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17599</th>\n",
       "      <td>site.02.subj.0910.lab.22A144.iso.1</td>\n",
       "      <td>LEV</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.00</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>Lineage 2</td>\n",
       "      <td>CHN</td>\n",
       "      <td>MDR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17600</th>\n",
       "      <td>site.05.subj.CA-0335.lab.CO-08338-18.iso.1</td>\n",
       "      <td>MXF</td>\n",
       "      <td>0.5</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>S</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>Lineage 4</td>\n",
       "      <td>PER</td>\n",
       "      <td>INH_AND_RIF_S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17601</th>\n",
       "      <td>site.02.subj.0016.lab.2014231005.iso.1</td>\n",
       "      <td>LEV</td>\n",
       "      <td>0.25</td>\n",
       "      <td>-2.00</td>\n",
       "      <td>S</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>Lineage 2</td>\n",
       "      <td>CHN</td>\n",
       "      <td>INH_AND_RIF_S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17602</th>\n",
       "      <td>site.05.subj.LS-1114.lab.MA-00010-19.iso.1</td>\n",
       "      <td>LEV</td>\n",
       "      <td>0.5</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>S</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>Lineage 4</td>\n",
       "      <td>PER</td>\n",
       "      <td>INH_AND_RIF_S</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17603 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          UNIQUEID DRUG     MIC  LOG2MIC  \\\n",
       "0       site.06.subj.06TB_1032.lab.06MIL2037.iso.1  MXF    0.25    -2.00   \n",
       "1       site.06.subj.06TB_0404.lab.06MIL1373.iso.1  MXF    0.25    -2.00   \n",
       "2      site.06.subj.SSM_0197-R.lab.06MIL0117.iso.1  MXF  <=0.06    -4.06   \n",
       "3              site.04.subj.01328.lab.722926.iso.1  LEV     0.5    -1.00   \n",
       "4               site.03.subj.T1069.lab.T1069.iso.1  LEV     0.5    -1.00   \n",
       "...                                            ...  ...     ...      ...   \n",
       "17598   site.05.subj.CA-1473.lab.CO-03323-19.iso.1  MXF     0.5    -1.00   \n",
       "17599           site.02.subj.0910.lab.22A144.iso.1  LEV     4.0     2.00   \n",
       "17600   site.05.subj.CA-0335.lab.CO-08338-18.iso.1  MXF     0.5    -1.00   \n",
       "17601       site.02.subj.0016.lab.2014231005.iso.1  LEV    0.25    -2.00   \n",
       "17602   site.05.subj.LS-1114.lab.MA-00010-19.iso.1  LEV     0.5    -1.00   \n",
       "\n",
       "      BINARY_PHENOTYPE PHENOTYPE_QUALITY LINEAGE_NAME  \\\n",
       "0                    S              HIGH    Lineage 4   \n",
       "1                    S              HIGH    Lineage 4   \n",
       "2                    S              HIGH    Lineage 3   \n",
       "3                    S              HIGH    Lineage 2   \n",
       "4                    S              HIGH    Lineage 4   \n",
       "...                ...               ...          ...   \n",
       "17598                S              HIGH    Lineage 4   \n",
       "17599                R              HIGH    Lineage 2   \n",
       "17600                S              HIGH    Lineage 4   \n",
       "17601                S              HIGH    Lineage 2   \n",
       "17602                S              HIGH    Lineage 4   \n",
       "\n",
       "      COUNTRY_WHERE_SAMPLE_TAKEN     BACKGROUND  \n",
       "0                            NaN      INH_MONOR  \n",
       "1                            NaN  INH_AND_RIF_S  \n",
       "2                            PAK  INH_AND_RIF_S  \n",
       "3                            IND  INH_AND_RIF_S  \n",
       "4                            TJK      RIF_MONOR  \n",
       "...                          ...            ...  \n",
       "17598                        PER  INH_AND_RIF_S  \n",
       "17599                        CHN            MDR  \n",
       "17600                        PER  INH_AND_RIF_S  \n",
       "17601                        CHN  INH_AND_RIF_S  \n",
       "17602                        PER  INH_AND_RIF_S  \n",
       "\n",
       "[17603 rows x 9 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#merge files to give df with sample uniqueid, lineage, country of origin and phenotypes to LEV/MXF\n",
    "gs=pd.merge(genomes, samples, how='left', on=['SITEID','SUBJID','LABID'])\n",
    "gs = gs[['UNIQUEID','LINEAGE_NAME','COUNTRY_WHERE_SAMPLE_TAKEN']]\n",
    "gs=pd.merge(gs, background_df, how='left', on=['UNIQUEID'])\n",
    "pgs=pd.merge(fq_phenotypes, gs, how='left', on=['UNIQUEID'])\n",
    "pgs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "#merge to get mutations in each isolate and remove any isolates with no MIC\n",
    "pgsm = pd.merge(pgs, gyrase_mutations, how='left', on='UNIQUEID')\n",
    "mic_nan=pgsm.loc[pgsm['MIC'].isna()].index\n",
    "pgsm=pgsm.drop(mic_nan)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UNIQUEID</th>\n",
       "      <th>DRUG</th>\n",
       "      <th>MIC</th>\n",
       "      <th>LOG2MIC</th>\n",
       "      <th>BINARY_PHENOTYPE</th>\n",
       "      <th>PHENOTYPE_QUALITY</th>\n",
       "      <th>LINEAGE_NAME</th>\n",
       "      <th>COUNTRY_WHERE_SAMPLE_TAKEN</th>\n",
       "      <th>BACKGROUND</th>\n",
       "      <th>GENE</th>\n",
       "      <th>MUTATION</th>\n",
       "      <th>POSITION</th>\n",
       "      <th>GENEMUT</th>\n",
       "      <th>IS_HET</th>\n",
       "      <th>NUMBER_OF_MUTATIONS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>site.06.subj.06TB_1032.lab.06MIL2037.iso.1</td>\n",
       "      <td>MXF</td>\n",
       "      <td>0.25</td>\n",
       "      <td>-2.00</td>\n",
       "      <td>S</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>Lineage 4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>INH_MONOR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>site.06.subj.06TB_0404.lab.06MIL1373.iso.1</td>\n",
       "      <td>MXF</td>\n",
       "      <td>0.25</td>\n",
       "      <td>-2.00</td>\n",
       "      <td>S</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>Lineage 4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>INH_AND_RIF_S</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>site.06.subj.SSM_0197-R.lab.06MIL0117.iso.1</td>\n",
       "      <td>MXF</td>\n",
       "      <td>&lt;=0.06</td>\n",
       "      <td>-4.06</td>\n",
       "      <td>S</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>Lineage 3</td>\n",
       "      <td>PAK</td>\n",
       "      <td>INH_AND_RIF_S</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>site.04.subj.01328.lab.722926.iso.1</td>\n",
       "      <td>LEV</td>\n",
       "      <td>0.5</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>S</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>Lineage 2</td>\n",
       "      <td>IND</td>\n",
       "      <td>INH_AND_RIF_S</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>site.03.subj.T1069.lab.T1069.iso.1</td>\n",
       "      <td>LEV</td>\n",
       "      <td>0.5</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>S</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>Lineage 4</td>\n",
       "      <td>TJK</td>\n",
       "      <td>RIF_MONOR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18767</th>\n",
       "      <td>site.05.subj.CA-1473.lab.CO-03323-19.iso.1</td>\n",
       "      <td>MXF</td>\n",
       "      <td>0.5</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>S</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>Lineage 4</td>\n",
       "      <td>PER</td>\n",
       "      <td>INH_AND_RIF_S</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18768</th>\n",
       "      <td>site.02.subj.0910.lab.22A144.iso.1</td>\n",
       "      <td>LEV</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.00</td>\n",
       "      <td>R</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>Lineage 2</td>\n",
       "      <td>CHN</td>\n",
       "      <td>MDR</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>A90V</td>\n",
       "      <td>90.0</td>\n",
       "      <td>gyrAA90V</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18769</th>\n",
       "      <td>site.05.subj.CA-0335.lab.CO-08338-18.iso.1</td>\n",
       "      <td>MXF</td>\n",
       "      <td>0.5</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>S</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>Lineage 4</td>\n",
       "      <td>PER</td>\n",
       "      <td>INH_AND_RIF_S</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18770</th>\n",
       "      <td>site.02.subj.0016.lab.2014231005.iso.1</td>\n",
       "      <td>LEV</td>\n",
       "      <td>0.25</td>\n",
       "      <td>-2.00</td>\n",
       "      <td>S</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>Lineage 2</td>\n",
       "      <td>CHN</td>\n",
       "      <td>INH_AND_RIF_S</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18771</th>\n",
       "      <td>site.05.subj.LS-1114.lab.MA-00010-19.iso.1</td>\n",
       "      <td>LEV</td>\n",
       "      <td>0.5</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>S</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>Lineage 4</td>\n",
       "      <td>PER</td>\n",
       "      <td>INH_AND_RIF_S</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18412 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          UNIQUEID DRUG     MIC  LOG2MIC  \\\n",
       "0       site.06.subj.06TB_1032.lab.06MIL2037.iso.1  MXF    0.25    -2.00   \n",
       "1       site.06.subj.06TB_0404.lab.06MIL1373.iso.1  MXF    0.25    -2.00   \n",
       "2      site.06.subj.SSM_0197-R.lab.06MIL0117.iso.1  MXF  <=0.06    -4.06   \n",
       "3              site.04.subj.01328.lab.722926.iso.1  LEV     0.5    -1.00   \n",
       "4               site.03.subj.T1069.lab.T1069.iso.1  LEV     0.5    -1.00   \n",
       "...                                            ...  ...     ...      ...   \n",
       "18767   site.05.subj.CA-1473.lab.CO-03323-19.iso.1  MXF     0.5    -1.00   \n",
       "18768           site.02.subj.0910.lab.22A144.iso.1  LEV     4.0     2.00   \n",
       "18769   site.05.subj.CA-0335.lab.CO-08338-18.iso.1  MXF     0.5    -1.00   \n",
       "18770       site.02.subj.0016.lab.2014231005.iso.1  LEV    0.25    -2.00   \n",
       "18771   site.05.subj.LS-1114.lab.MA-00010-19.iso.1  LEV     0.5    -1.00   \n",
       "\n",
       "      BINARY_PHENOTYPE PHENOTYPE_QUALITY LINEAGE_NAME  \\\n",
       "0                    S              HIGH    Lineage 4   \n",
       "1                    S              HIGH    Lineage 4   \n",
       "2                    S              HIGH    Lineage 3   \n",
       "3                    S              HIGH    Lineage 2   \n",
       "4                    S              HIGH    Lineage 4   \n",
       "...                ...               ...          ...   \n",
       "18767                S              HIGH    Lineage 4   \n",
       "18768                R              HIGH    Lineage 2   \n",
       "18769                S              HIGH    Lineage 4   \n",
       "18770                S              HIGH    Lineage 2   \n",
       "18771                S              HIGH    Lineage 4   \n",
       "\n",
       "      COUNTRY_WHERE_SAMPLE_TAKEN     BACKGROUND  GENE MUTATION  POSITION  \\\n",
       "0                            NaN      INH_MONOR   NaN      NaN       NaN   \n",
       "1                            NaN  INH_AND_RIF_S   NaN      NaN       NaN   \n",
       "2                            PAK  INH_AND_RIF_S   NaN      NaN       NaN   \n",
       "3                            IND  INH_AND_RIF_S   NaN      NaN       NaN   \n",
       "4                            TJK      RIF_MONOR   NaN      NaN       NaN   \n",
       "...                          ...            ...   ...      ...       ...   \n",
       "18767                        PER  INH_AND_RIF_S   NaN      NaN       NaN   \n",
       "18768                        CHN            MDR  gyrA     A90V      90.0   \n",
       "18769                        PER  INH_AND_RIF_S   NaN      NaN       NaN   \n",
       "18770                        CHN  INH_AND_RIF_S   NaN      NaN       NaN   \n",
       "18771                        PER  INH_AND_RIF_S   NaN      NaN       NaN   \n",
       "\n",
       "        GENEMUT  IS_HET  NUMBER_OF_MUTATIONS  \n",
       "0           NaN     NaN                    0  \n",
       "1           NaN     NaN                    0  \n",
       "2           NaN     NaN                    0  \n",
       "3           NaN     NaN                    0  \n",
       "4           NaN     NaN                    0  \n",
       "...         ...     ...                  ...  \n",
       "18767       NaN     NaN                    0  \n",
       "18768  gyrAA90V     0.0                    1  \n",
       "18769       NaN     NaN                    0  \n",
       "18770       NaN     NaN                    0  \n",
       "18771       NaN     NaN                    0  \n",
       "\n",
       "[18412 rows x 15 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#count number of dna gyrase mutations per sample\n",
    "no_muts=[]\n",
    "for index, row in pgsm.iterrows():\n",
    "    no_muts.append(pgsm.loc[pgsm.UNIQUEID==row['UNIQUEID']].GENEMUT.nunique(dropna=True))\n",
    "pgsm['NUMBER_OF_MUTATIONS']=no_muts\n",
    "pgsm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Int64Index: 18412 entries, 0 to 18411\n",
      "Data columns (total 52 columns):\n",
      " #   Column                           Non-Null Count  Dtype   \n",
      "---  ------                           --------------  -----   \n",
      " 0   UNIQUEID                         18412 non-null  object  \n",
      " 1   DRUG                             18412 non-null  category\n",
      " 2   MIC                              18412 non-null  category\n",
      " 3   LOG2MIC                          18412 non-null  float64 \n",
      " 4   BINARY_PHENOTYPE                 18412 non-null  category\n",
      " 5   PHENOTYPE_QUALITY                18412 non-null  category\n",
      " 6   LINEAGE_NAME                     18412 non-null  category\n",
      " 7   COUNTRY_WHERE_SAMPLE_TAKEN       15147 non-null  category\n",
      " 8   BACKGROUND                       18412 non-null  object  \n",
      " 9   GENE                             6391 non-null   object  \n",
      " 10  MUTATION                         6391 non-null   object  \n",
      " 11  POSITION_x                       6391 non-null   float64 \n",
      " 12  GENEMUT                          6391 non-null   object  \n",
      " 13  IS_HET                           6391 non-null   float64 \n",
      " 14  NUMBER_OF_MUTATIONS              18412 non-null  int64   \n",
      " 15  OLD_AA                           6391 non-null   object  \n",
      " 16  POSITION_y                       6391 non-null   float64 \n",
      " 17  NEW_AA                           6391 non-null   object  \n",
      " 18  ATOM_CHANGE                      6391 non-null   float64 \n",
      " 19  VOLUME_CHANGE                    6391 non-null   float64 \n",
      " 20  HYDROPATHY_CHANGE                6391 non-null   float64 \n",
      " 21  CHARGE_CHANGE                    6391 non-null   float64 \n",
      " 22  HDONOR_CHANGE                    6391 non-null   float64 \n",
      " 23  HACCEPTOR_CHANGE                 6391 non-null   float64 \n",
      " 24  ROTABLE_BOND_CHANGE              6391 non-null   float64 \n",
      " 25  AROMATIC_RING_CHANGE             6391 non-null   float64 \n",
      " 26  SULPHUR_CHANGE                   6391 non-null   float64 \n",
      " 27  MAPP_SCORE                       6391 non-null   float64 \n",
      " 28  RESIDUE_DEPTH                    6391 non-null   float64 \n",
      " 29  LIGAND_DISTANCE                  6391 non-null   float64 \n",
      " 30  LIGAND_MG_DISTANCE               6391 non-null   float64 \n",
      " 31  CATALYTIC_MG_DISTANCE            6391 non-null   float64 \n",
      " 32  DNA_DISTANCE                     6391 non-null   float64 \n",
      " 33  RSA                              6391 non-null   float64 \n",
      " 34  PP_AFFINITY_CHANGE               6391 non-null   float64 \n",
      " 35  STABILITY_CHANGE                 6391 non-null   float64 \n",
      " 36  PROTEIN_HBONDS                   6391 non-null   float64 \n",
      " 37  DNA_HBONDS                       6391 non-null   float64 \n",
      " 38  PTR_HBONDS                       6391 non-null   float64 \n",
      " 39  LIGAND_HBONDS                    6391 non-null   float64 \n",
      " 40  SECONDARY_STRUCTURE_B            6391 non-null   float64 \n",
      " 41  SECONDARY_STRUCTURE_E            6391 non-null   float64 \n",
      " 42  SECONDARY_STRUCTURE_G            6391 non-null   float64 \n",
      " 43  SECONDARY_STRUCTURE_H            6391 non-null   float64 \n",
      " 44  SECONDARY_STRUCTURE_I            6391 non-null   float64 \n",
      " 45  SECONDARY_STRUCTURE_NaN          6391 non-null   float64 \n",
      " 46  SECONDARY_STRUCTURE_S            6391 non-null   float64 \n",
      " 47  SECONDARY_STRUCTURE_T            6391 non-null   float64 \n",
      " 48  LIGAND_INTERACTION_NaN           6391 non-null   float64 \n",
      " 49  LIGAND_INTERACTION_none          6391 non-null   float64 \n",
      " 50  LIGAND_INTERACTION_salt_bridge   6391 non-null   float64 \n",
      " 51  LIGAND_INTERACTION_water_bridge  6391 non-null   float64 \n",
      "dtypes: category(6), float64(38), int64(1), object(7)\n",
      "memory usage: 6.7+ MB\n"
     ]
    }
   ],
   "source": [
    "#merge with structure based features for ML\n",
    "final_df = pd.merge(pgsm, features, how='left', on=['GENE','MUTATION'])\n",
    "final_df.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "#rename columns\n",
    "final_df=final_df.rename(columns={\"POSITION_x\": \"POSITION\"})\n",
    "final_df=final_df.rename(columns={\"MAPP_SCORE\": \"SNAP2_SCORE\"})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "final_df.to_csv('ML_DATA_HIGHQ.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
